feat(checkhealth): support Lua healthchecks

muniter opened the pull request:
# Lua Healthchecks
This PR does:

- Refactor health.vim to discover lua healthcheck in the runtime files `lua/**/health{/init}.lua`
- Support healthchecks for lua submodules e.g `:checkhealth vim.lsp` and also support wildcard `*` at the end to discover all submodules `:checkhealth vim*`
- Refactor health.vim to use variable scope instead of output capturing Create **health.lua** module to wrap report functions and future extensibility.
- Add tests to cover all the new functionality
- Update the documentation to reflect the changes.
- Refactor completion for the ex command to provide lua module completions.

## Missing
- [x] Ex command completion for lua submodules<del>, tried looking at the C code, **at the moment too complicated for me to figure it out**. Don't see an easy way to provide a lua or  vimscript for it instead.</del>

## Example:
Plugin linter in rtp can declare it's checkhealts in lua module `lua/linter/health{/init}.lua`,  that returns a table with a method "check" that when executed calls the report functions provided by the builtin lua module `require("health")`.

The plugin also has a submodule `lua/linter/providers` in which it defines `lua/linter/providers/health{/init}.lua`

This plugin healthcheck can now be run by the ex command:
*  `:checkhealth linter` for the base module.
* `:checkhealth  linter.providers` for the providers submodule.

Also checking all submodules can be done by:
* `:checkhealth linter*` -> which expands to `:checkhealth linter linter.providers`

The base  and any submodules will be discovered by the standard:
* `:checkhealth`

## Other notes

- With this PR the **lua healthchecks written for [treesitter](https://github.com/neovim/neovim/blob/5e22fdd9ccdd894d4c6e165438ca314ed2cd3b9e/runtime/lua/vim/treesitter/health.lua) and [lsp](https://github.com/neovim/neovim/blob/5e22fdd9ccdd894d4c6e165438ca314ed2cd3b9e/runtime/lua/vim/lsp/health.lua) would work out of the box**, only requiring to change their function `check_health()` for `check()` and this way removing the autoload wrappers in `runtime/autoload/health`.
- Updates the plugin to use `get_runtime_files` instead of `globpath(&rtp ...` due to #15632 

@tjdevries you told me to ping you 😄 

Fixes #15099

muniter committed:
feat(runtime/health): support lua healthchecks

- Refactor health.vim to discover lua healthcheck in the runtime
  directories lua/**/health{/init}.lua
- Support healthchecks for lua submodules e.g :checkhealth vim.lsp and
  also support wildcard "*" at the end for all submodules
  :checkhealth vim*
- Refactor health.vim to use variable scope instead of output capturing
- Create health.lua module to wrap report functions and future
  extensibility.
- Move away from searching just in the runtimepath, use
  `nvim_get_runtime_file` due to #15632

Example:
Plugin linter in rtp can declare it's checkhealts in lua module
`lua/linter/health{/init}.lua` that returns a table with a method
"check" that when executed calls the report functions provided by the
builtin lua module require("health").

The plugin also has a submodule `/lua/linter/providers` in which it
defines `/lua/linter/providers/health{/init}.lua`

This plugin healthcheck can now be run by the ex command:
  `:checkhealth linter linter.providers`

Also calling all submodules can be done by:
  `:checkhealth linter*

And "linter" and "linter.provider" would be discovered when:
  `:checkhealth`

muniter committed:
test(runtime/health): cover lua healthchecks

- Add tests for lua healthchecks (failure, success and submodules).
- Reword some of the test naming for improved logs readability.
- Modify render test to accomodate the changes of the health autoload function.
- Add test for :checkhealth completion of Lua healtchecks.

muniter commented:
I'm open to use the `autoload/health/plugin.lua` if it's a better fit. Thanks for linking the feature request.

justinmk commented:
@muniter try a squash-merge or just a merge instead of a rebase. unless you have commits you want to keep separated in this PR.

muniter commented:
Thanks @justinmk I've reverted and applied my changes!

bfredl commented on test/functional/plugin/health_spec.lua:
```diff
@@ -80,36 +80,72 @@ describe('health.vim', function()
         ## another 1
           - OK: ok
+
+        lua_success: v:lua.require("health").check("test_plug.health.lua_success")
```
this a little weird, I don't think lua checks should be marked as "secondary" to vimL functions. perhaps just `require("test_plug.health.lua_success")` which is equivalent of the autoload file references (which not are complete calls).

justinmk commented on runtime/lua/health.lua:
```diff
@@ -0,0 +1,15 @@
+  return function()
+    if not ok then
+      -- print instead of return since the output of functions is what's being
+      -- captured not it's return value
```
can we somehow avoid perpetuating this design mistake in the Lua impl?

justinmk commented on test/functional/fixtures/lua/test_plug/health/lua_success.lua:
```diff
@@ -0,0 +1,10 @@
+  vim.fn['health#report_start']("report 1")
+  vim.fn['health#report_ok']("everything is fine")
+  vim.fn['health#report_start']("report 2")
+  vim.fn['health#report_ok']("nothing to see here")
```
we might as well take this opportunity to introduce a `lua/health.lua` module with proper Lua `health.report_xx()` functions. They can just wrap `vim.fn['...']` for now, but let's not perpetuate that in the Lua healthchecks themselves.

justinmk commented on test/functional/plugin/health_spec.lua:
```diff
@@ -126,23 +167,24 @@ describe('health.vim', function()
     end)
 
-    it("gracefully handles invalid healthcheck", function()
+    it("#ihc gracefully handles invalid healthcheck", function()
```
revert

justinmk commented on test/functional/plugin/health_spec.lua:
```diff
@@ -80,36 +81,76 @@ describe('health.vim', function()
+        ]])
+    end)
+
+    it("checkealths lua plugins", function()
```
```suggestion
    it(":checkhealth lua plugins", function()
```

justinmk commented on runtime/autoload/health.vim:
```diff
@@ -147,31 +151,41 @@ function! health#report_error(msg, ...) abort " {{{
-" Translates a list of plugin names to healthcheck function names.
-function! s:to_fn_names(plugin_names) abort
+" Return a list of [{name}, {callable}, {display_name}].
+function! s:get_health_check(plugin_names) abort
```
use existing conventions, do not separate works randomly.
```suggestion
function! s:get_healthcheck(plugin_names) abort
```

justinmk commented on runtime/autoload/health.vim:
```diff
@@ -43,19 +43,23 @@ function! health#check(plugin_names) abort
-        let output = "\n\n".execute('call '.c.'()')
+        let output = "\n\n".execute('call '.c[1].'()')
+        if output =~# 'HEALTHCHECK_ERROR:'
+          throw matchlist(output, 'HEALTHCHECK_ERROR:\(.*$\)')[1]
```
instead of this , let's require Lua healthchecks to return a structure (string, or even a map of some sort) instead of this output scraping. The output scraping was a design mistake.

muniter replied:
I refactored health.vim to use variable scope, now every report function instead of echoing the formatted parameters appends to an `output` list. So now in the lua side things are cleaner the report functions are wrapped to look nicer. Let me know what you think.

justinmk commented on runtime/autoload/health.vim:
```diff
@@ -42,24 +42,26 @@ function! health#check(plugin_names) abort
       try
-        let output = "\n\n".execute('call '.c.'()')
+        if c[1] == ''
+          throw 'healthcheck not found'
```
`c[1]` is already used above, why do this check here? also weird to throw in this try-catch

justinmk commented on runtime/autoload/health.vim:
```diff
@@ -71,9 +73,13 @@ function! health#check(plugin_names) abort
   redraw|echo ''
 endfunction
 
+function! health#output(output)
```
doesn't look like this needs to be public.
```suggestion
function! s:collect_output(output)
```

justinmk commented on test/functional/plugin/health_spec.lua:
```diff
@@ -48,68 +49,78 @@ describe('health.vim', function()
-  it("health#report_*()", function()
-    helpers.source([[
-      let g:health_report = execute([
-        \ "call health#report_start('Check Bar')",
```
does it make sense to keep this test, but add a `health#render()` or something, that returns the formatted final result?

muniter replied:
I don't see the need, we are already testing all the functions render in the other functional tests that do `:chackealth sucess broken`

muniter committed:
docs(runtime/health): update with new lua support

- describe how the lua support works
- explain new behavior of :checkhealth *
- fixed formatting to use tab instead of spaces

muniter commented:
I've made some good progress and updated the docs for clarity and added a nice feature I like of health check also in "submodules" which is something we do already with `vim.treesitter` and `vim.lsp` I like the "*" wildcard feature. I hope to hear your comments 😄

muniter commented on runtime/autoload/health.vim:
```diff
@@ -42,24 +42,29 @@ function! health#check(plugin_names) abort
+        let s:output = []  " Clear the output
+        if v:exception =~# 'healthcheck_not_found'
+          call health#report_error('No healthcheck found for "'.name.'" plugin.')
+          let func = ""  " don't display, unknown if healthcheck is lua or vimscrip
```
This can be deleted, only way to get here already checks `func ==`

muniter commented on test/functional/plugin/health_spec.lua:
```diff
@@ -48,68 +49,144 @@ describe('health.vim', function()
-      ## Baz
-        - WARNING: Zim
+  describe(":checkhealth", function()
+    it("functions health#report_*() render correctly", function()
```
@justinmk went ahead and created this test case to render a checkhealth that uses all the function features (advice, help docs references) to compensate for the previous one

gpanders commented on runtime/doc/pi_health.txt:
```diff
@@ -35,15 +35,98 @@ Commands				*health-commands*
+==============================================================================
+Functions				*health-functions-lua*
+
+health.vim lua functions are for creating new healthchecks. They mostly just
```
Treat `Lua` as a proper noun (i.e. capitalize the first letter).

gpanders commented on runtime/doc/pi_health.txt:
```diff
@@ -35,15 +35,98 @@ Commands				*health-commands*
+
+Lua					|health-lua|
+
+The lua health module provides the functions:
+
+  `local health = require('health')`
+
```
Literal blocks in help syntax are used like this:

```suggestion
The lua health module provides the functions: >
    local health = require('health')
<
```

gpanders commented on runtime/doc/pi_health.txt:
```diff
@@ -35,15 +35,98 @@ Commands				*health-commands*
+
+health.report_start({name})				*health.report_start()*
+	Starts a new report. Most plugins should call this only once, but if
+        you want different sections to appear in your report, call this once
+        per section.
```
Use hard tabs for indentation in help docs (the help file should have a modeline already with something like `noet ts=8 sw=0 tw=78 ft=help`, feel free to add it if it doesn't exist)

muniter replied:
I went ahead and changed all the indentation for tabs! thanks for the review.

gpanders commented on runtime/doc/pi_health.txt:
```diff
@@ -35,15 +35,98 @@ Commands				*health-commands*
+health.report_ok({msg})					*health.report_ok()*
+	Reports a "success" message.
+
+health.report_warn({msg}, [{advice}])			*health.report_warn()*
```
For optional arguments include the `,` in the brackets:
```suggestion
health.report_warn({msg}[, {advice}])			*health.report_warn()*
```

gpanders commented on runtime/doc/pi_health.txt:
```diff
@@ -13,7 +13,7 @@ python support, ruby support, clipboard support, and more.
 To run the healthchecks, use this command: >
 
-        :checkhealth
+:checkhealth
```
Why is the indentation removed?

gpanders commented on runtime/doc/pi_health.txt:
```diff
@@ -23,60 +23,135 @@ Commands				*health-commands*
+==============================================================================
+Lua Functions				*health-functions-lua* *health-lua*
+
+health.vim Lua functions are for creating new healthchecks. They mostly just
+do some layout and formatting, to give users a consistent presentation. See
+also |health-functions-vim|
+
+The Lua health module provides the functions:  >
+	local health = require('health')
+<
```
I don't think you need the "mostly just do some layout and formatting" part. If the role of the functions changes in the future then this becomes stale.

```suggestion
The Lua "health" module can be used to create new healthchecks (see also
|health-functions-vim|). To get started, simply use:  >
	local health = require('health')
<
```

gpanders commented on runtime/doc/pi_health.txt:
```diff
@@ -23,60 +23,135 @@ Commands				*health-commands*
+Healthchecks are functions that check the user environment, configuration,
+etc. Nvim has built-in healthchecks in $VIMRUNTIME/autoload/health/.
+
+To add a new healthcheck for your own plugin, simply define a Lua module in
+your plugin that returns a table with a function check(), |:checkhealth|
+automatically finds and invokes such modules function.
```
```suggestion
To add a new healthcheck for your own plugin, simply define a Lua module in
your plugin that returns a table with a "check()" function. |:checkhealth|
will automatically find and invoke this function.
```

gpanders commented on runtime/doc/pi_health.txt:
```diff
@@ -23,60 +23,135 @@ Commands				*health-commands*
+automatically finds and invokes such modules function.
+
+If your plugin is named "foo", then its healthcheck module should be a file in
+one of this locations on the 'runtimepath':
```
```suggestion
one of these locations on 'runtimepath':
```

gpanders commented on runtime/doc/pi_health.txt:
```diff
@@ -23,60 +23,135 @@ Commands				*health-commands*
+	- `lua/foo/health/init.lua`
+	- `lua/foo/health.lua`
+
+If your plugin provides a submodule named "bar" and you want to have a separate
+checkhealth for it you can define the healthcheck at one of this locations also
+on the 'runtimepath':
```
```suggestion
If your plugin provides a submodule named "bar" for which you want a separate
healthcheck, define the healthcheck at one of these locations on 'runtimepath':
```

gpanders commented on runtime/doc/pi_health.txt:
```diff
@@ -23,60 +23,135 @@ Commands				*health-commands*
+	- `lua/foo/bar/health/init.lua`
+	- `lua/foo/bar/health.lua`
+
+All submodules should return a Lua table containing the method `check()`
```
```suggestion
All submodules should return a Lua table containing the method `check()`.
```

gpanders commented on runtime/doc/pi_health.txt:
```diff
@@ -23,60 +23,135 @@ Commands				*health-commands*
+
+All submodules should return a Lua table containing the method `check()`
+
+Copy this sample code into `lua/foo/health/init.lua` or `lua/foo/health.lua`
+replacing "foo" in the path with your plugin name: >
```
```suggestion
Copy this sample code into `lua/foo/health/init.lua` or `lua/foo/health.lua`,
replacing "foo" in the path with your plugin name: >
```

gpanders commented on runtime/doc/pi_health.txt:
```diff
@@ -23,60 +23,135 @@ Commands				*health-commands*
+If your plugin is named "foo", then its healthcheck module should be a file in
+one of this locations on the 'runtimepath':
+	- `lua/foo/health/init.lua`
+	- `lua/foo/health.lua`
+
+If your plugin provides a submodule named "bar" and you want to have a separate
+checkhealth for it you can define the healthcheck at one of this locations also
+on the 'runtimepath':
+	- `lua/foo/bar/health/init.lua`
+	- `lua/foo/bar/health.lua`
+
+All submodules should return a Lua table containing the method `check()`
+
+Copy this sample code into `lua/foo/health/init.lua` or `lua/foo/health.lua`
+replacing "foo" in the path with your plugin name: >
+
+	local M = {}
+	local health = require("health")
+
+	M.check = function()
+	  health.report_start("my_plugin report")
+	  -- make sure setup function parameters are ok
+	  if check_setup() then
+	    health.report_ok("Setup function is correct")
+	  else
+	    health.report_error("Setup function is incorrect")
+	  end
+	  -- do some more checking
+	  -- ...
+	end
+
+	return M
+
+==============================================================================
+Vimscript Functions		*health-functions-vimscript* *health-vimscript*
 
 health.vim functions are for creating new healthchecks. They mostly just do
-some layout and formatting, to give users a consistent presentation.
+some layout and formatting, to give users a consistent presentation. See also
+|health-functions-lua|
```
Same comment as above, I don't think we need to mention that these functions just do layout and formatting. It's not relevant to the user.

gpanders reviewed (changes requested):
I just had some comments on the prose in the docs, the code LGTM!

muniter committed:
feat(ex_checkhealth): provide function for command line completion

Move away from providing completion with ExpandRTDir to ExpandGeneric
providing the function get_healthcheck_name which caches the results for
the current command line prompt.

It does the almost the same thing the Vim function 'get_healthcheck'
implemented in 'runtime/autoload/health.vim' does.

teto commented on src/nvim/ex_getln.c:
```diff
@@ -230,7 +230,22 @@ static int compl_match_arraysize;
 static int compl_startcol;
 static int compl_selected;
 
+/// Healtchcheck completion items
```
typo

teto commented:
would you mind rebasing and check linting eventually ? looks good to go

muniter commented:
Thanks for the review @teto I've rebased, renamed some variables to match the style guide, and fixed linting.

Currently the one job that failed was because it failed downloading LLVM.

teto commented:
cc @tjdevries @vigoux seems like this will break several healthchecks but the fix should be straightforward: add a check() function to your plugin health module:
```
24 nvim-treesitter: require("nvim-treesitter.health").check()                                                                                                                                                                                                                                                                  
 23 ========================================================================                                                                                                                                                                                                                                                    
 22 | - ERROR: Failed to run healthcheck for "nvim-treesitter" plugin. Exception:                                                                                                                                                                                                                                               
 21 |   function health#check, line 24                                                                                                                                                                                                                                                                                          
 20 |   Vim(eval):E5108: Error executing lua [string "luaeval()"]:1: attempt to call field 'check' (a nil value)                                                                                                                                                                                                                
 19                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        
  9                                                                                                                                                                                                                                                                                                                             
  8 ## ===== Installed extensions =====                                                                                                                                                                                                                                                                                         
  7                                                                                                                                                                                                                                                                                                                             
  6 ## Telescope Extension: `fzy_native`                                                                                                                                                                                                                                                                                        
  5 | - INFO: No healthcheck provided                                                                                                                                                                                                                                                                                           
  4                                                                                                                                                                                                                                                                                                                             
  3 telescope: require("telescope.health").check()                                                                                                                                                                                                                                                                              
  2 ========================================================================                                                                                                                                                                                                                                                    
  1 | - ERROR: Failed to run healthcheck for "telescope" plugin. Exception:                                                                                                                                                                                                                                                     
235 |   function health#check, line 24                                                                                                                                                                                                                                                                                          
  1 |   Vim(eval):E5108: Error executing lua [string "luaeval()"]:1: attempt to call field 'check' (a nil value)             
```

clason commented:
also lspconfig @neovim/lsp

clason commented:
How is the situation handled if you have both Lua `check()` and the `autoload` shim (you need both to support stable versions of Neovim)? Are plugins supposed to `has('nvim-0.6')` the shim, or is this implementation smart enough to skip the shim if the Lua function exists?

clason commented:
Also the builtin `vim.lsp.health` and `vim.treesitter.health` need to be updated!

teto commented:
plugins need the `has('nvim-0.6')` otherwise it runs twice, but even I think it will still try to call health#nvim#check so you will have the header twice.

teto commented:
seems like all plugins use M.check_health so maybe we should just use that instead of just M.check

clason commented:
I think it would be better to handle this on the level of the healthcheck function, then? (Try `require` first, and only if that fails, call `autoload#health#check`.)

clason commented:
> seems like all plugins use M.check_health so maybe we should just use that instead of just M.check

Or at least have this as a deprecation:

1. try `health.check()`
2. if that fails, try `health.check_health()`; print a deprecation warning if it succeeds
3. if that fails, try `autoload#health#check`

clason commented on runtime/autoload/health.vim:
```diff
@@ -119,59 +127,74 @@ endfunction " }}}
+          \ + nvim_get_runtime_file('lua/**/'.p.'/health/init.lua', v:true)
+          \ + nvim_get_runtime_file('lua/**/'.p.'/health.lua', v:true)
+    if len(paths) == 0
+      let healthchecks += [[p, '', '']]  " healthchek not found
```
typo: `healthchek`

clason commented on src/nvim/ex_getln.c:
```diff
@@ -273,6 +288,68 @@ static void init_incsearch_state(incsearch_state_T *s)
+         (int)healthchecks.names.ga_len ? ((char_u **)(healthchecks.names.ga_data))[idx] : NULL;
+}
+
+/// Transform healthcheck file path into it's name.
```
typo: `it's` -> `its`

muniter commented:
@teto @clason I did mention it the last section of the main post, I'm very sorry If I didn't make it clear enough, or warned.

> With this PR the lua healthchecks written for treesitter and lsp would work out of the box, only requiring to change their function check_health() for check() and this way removing the autoload wrappers in runtime/autoload/health.

I think the way forward with https://github.com/neovim/nvim-lspconfig/pull/1296 should be to just change it here in core, and I can send a PR to the plugins I know of. I went with `health` instead of `health_check` to avoid the redundant use of the word, in the headings

`nvim-treesitter: require("nvim-treesitter.health").check()`

vs

`nvim-treesitter: require("nvim-treesitter.health").health_check()`

I think the former is easier to read, no double health, also the module already is related to health.

https://github.com/neovim/neovim/pull/15914 Fixes the issue for the core healthchecks

clason commented:
I agree that this should be changed in core, especially to avoid the duplicate health check. The new name is better, but I'm not sure it's worth the sudden breakage? 

I still think a strategy like https://github.com/neovim/neovim/pull/15259#issuecomment-934248559 would be the best approach here.

clason commented:
Alternative is to backport this to 0.5.2, get this out of the door quickly, and rip the band-aid off all affected plugins.

justinmk commented:
> The new name is better, but I'm not sure it's worth the sudden breakage?

what breakage?  this is a new feature.

clason commented:
> what breakage? this is a new feature.

Not quite -- the implementation is "eager": if there is a `lua/health.lua`, it will look for a function `check()` inside and print an error message if it can't find it. This _is_ the case for most plugins that use the `autoload#health#check` function to call `lua require'health'.check_health()` (note the name):
https://github.com/neovim/nvim-lspconfig/blob/master/autoload/health/lspconfig.vim

This pattern was introduced in core and copied throughout the ecosystem.

justinmk commented on runtime/doc/pi_health.txt:
```diff
@@ -23,60 +23,131 @@ Commands				*health-commands*
+		find the standard "runtime files" for syntax highlighting,
+		filetype-specific behavior, and standard plugins (including
+		:checkhealth).  If the runtime files cannot be found then
+		those features will not work.
```
please do NOT convert Nvim-owned helpfiles to tabs.

justinmk commented on runtime/doc/pi_health.txt:
```diff
@@ -23,60 +23,131 @@ Commands				*health-commands*
+If your plugin is named "foo", then its healthcheck module should be a file in
+one of these locations on 'runtimepath' or 'packpath':
+	- `lua/foo/health/init.lua`
+	- `lua/foo/health.lua`
```
do not over-use backticks. They make it less convenient to copy-paste the names. Also the highlighting draws attention to stuff that often is not very important. Why highlight these filenames?
